ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -788,6 +833,286 @@ private void migrateToPropertyStore() throws InterruptedException {
+                Collections.emptyMap());
+
+            // 6. Update ideal state and enable waged rebalancer
+            for (String resource : resources) {
+              IdealState idealState = resourceToIdealState.get(resource);
+              idealState.setRebalanceMode(IdealState.RebalanceMode.FULL_AUTO);
+              // Set resource tag
+              idealState.setInstanceGroupTag(getResourceTag(resource));
+              // Set number of replicas
+              idealState.setReplicas(NUM_REPLICAS);
+              idealState.setMinActiveReplicas(MIN_ACTIVE_REPLICAS);
+              helixAdmin.updateIdealState(clusterName, resource, idealState);
+            }
+            helixAdmin.enableWagedRebalance(clusterName, new ArrayList<>(resources));
+
+            // 7. Exit maintenance mode
+            helixAdmin.manuallyEnableMaintenanceMode(clusterName, false, "Complete migrating to Full auto",
+                Collections.emptyMap());
+
+            // After
+            Map<String, Set<String>> replicaPlacementAfter = getReplicaPlacement(helixAdmin, resources);
+
+            // Log stats
+            logMigrationStats(dcName, propertyStoreAdapter, replicaPlacementBefore, replicaPlacementAfter);
+          }
+          logger.info("Successfully migrated resources to full-auto in {}", dcName);
+        }
+      } catch (Throwable t) {
+        logger.error("Error while migrating resources to full-auto in {}", dcName, t);
+      } finally {
+        migrationComplete.countDown();
+      }
+    }, false).start());
+
+    migrationComplete.await();
+  }
+
+  /**
+   * Return tag for resources and their instances
+   * @param resource resource name
+   * @return tag name
+   */
+  private String getResourceTag(String resource) {
+    return "TAG_" + resource;
+  }
+
+  /**
+   * Get replica placement.
+   * @param helixAdmin {@link HelixAdmin} to talk to helix
+   * @param resources list of resources for which we need the partition-replica placement.
+   * @return
+   */
+  private Map<String, Set<String>> getReplicaPlacement(HelixAdmin helixAdmin, Set<String> resources) {
+    Map<String, Set<String>> partitionToInstances = new HashMap<>();
+    for (String resource : resources) {
+      // TODO: To check - In Helix full auto, ideal state may contain intermediate states.
+      IdealState idealState = helixAdmin.getResourceIdealState(clusterName, resource);
+      for (String partitionName : idealState.getPartitionSet()) {
+        partitionToInstances.put(partitionName, idealState.getInstanceSet(partitionName));
+      }
+    }
+    return partitionToInstances;
+  }
+
+  /**
+   * Sets up cluster to use waged rebalancer
+   * @param configAccessor the {@link ConfigAccessor} to access configuration in Helix.
+   */
+  private void setupCluster(ConfigAccessor configAccessor) {
+    ClusterConfig clusterConfig = configAccessor.getClusterConfig(clusterName);
```

Comment:
let's print out some log message before and after setting up cluster.

Reply:
Done